draft-ietf-tls-esni.md

```diff
@@ -521,57 +655,62 @@ PaddedServerNameList.sni and ClientESNIInner.nonce to the backend
 server. Thus, backend servers function the same as servers operating
 in Shared Mode.
 
+As in Shared Mode, if the backend server sends a NewSessionTicket message, the
+corresponding ESNI PSK MUST be ignored by other servers in the deployment when
+not negotiating ESNI, including servers which do not implement this
+specification.
+
 # Compatibility Issues
 
-In general, this mechanism is designed only to be used with
-servers which have opted in, thus minimizing compatibility
-issues. However, there are two scenarios where that does not
-apply, as detailed below.
+Unlike most TLS extensions, placing the SNI value in an ESNI extension
+is not interoperable with existing servers, which expect the value in
+the existing cleartext extension. Thus server operators SHOULD ensure
+servers understand a given set of ESNI keys before advertising them in
+DNS. Additionally, servers SHOULD retain support for any
+previously-advertised keys for the duration of their validity.
+
+However, in more complex deployment scenarios, this may be difficult
+to fully guarantee. Thus this protocol was designed to be robust in case
+of inconsistencies between DNS and servers, at the cost of extra
+round-trips due to a retry. Two specific scenarios are detailed below.
+
+## Misconfiguration and Deployment Concerns {#misconfiguration}
 
-## Misconfiguration
+It is possible for DNS and servers to become inconsistent. This may occur, for instance,
+from DNS misconfiguration, caching issues, or an incomplete rollout in a multi-server
+deployment. This may also occur if a server loses its ESNI keys, or an initial deployment
+of ESNI must be rolled back on the server.
 
-If DNS is misconfigured so that a client receives ESNI keys for a
-server which is not prepared to receive ESNI, then the server will
-ignore the "encrypted_server_name" extension, as required by
-{{RFC8446}}; Section 4.1.2.  If the servers does not
-require SNI, it will complete the handshake with its default
-certificate. Most likely, this will cause a certificate name
-mismatch and thus handshake failure. Clients SHOULD NOT fall
-back to cleartext SNI, because that allows a network attacker
-to disclose the SNI. They MAY attempt to use another server
-from the DNS results, if one is provided.
+The retry mechanism repairs most such inconsistencies. If server and DNS keys mismatch,
+the server will respond with esni_retry_requested. If the server does not understand the
+"encrypted_server_name" extension at all, it will ignore it as required by {{RFC8446}};
+Section 4.1.2. Provided the server can present a certificate valid for the public name,
+the client can safely retry with updated settings, as described in {{client-behavior}}.
 
+If the public name does not verify or the retry fails, the client SHOULD NOT
```

Comment:
Again, you need to separate retries that provide usable keys and ones that don't.

Reply:
I don't quite follow. What's the concern here? This one says "If the public name does not verify". Note this is also existing text, but slightly adapted. From before this PR:

> If the servers does not
> require SNI, it will complete the handshake with its default
> certificate. Most likely, this will cause a certificate name	
> mismatch and thus handshake failure. Clients SHOULD NOT fall	
> back to cleartext SNI, because that allows a network attacker	
> to disclose the SNI. They MAY attempt to use another server	
> from the DNS results, if one is provided.

Reply:
This is a matter of not leaning on other text properly.  There are three cases and this says two:

The public name can not verify, which is clearly a terminal error condition.  
The server can indicate that ESNI is disabled, which triggers a retry without ESNI.
The server can provide (usable) alternative keys, which triggers a retry with those keys.  If those keys are no good, that's a terminal condition.

I think that you can simplify this to say that unless ESNI is disabled as a result of successfully establishing a connection to the public name, then a client MUST NOT attempt to connect without ESNI.  (not SHOULD NOT)

Yes, this is probably not your fault.  I'll let others sort that bit out.

Reply:
Adopted your text with the MUST NOT.